source/custom/visa_service.custom.cpp

```diff
@@ -308,62 +571,57 @@ ::grpc::Status VisaService::ReadAsync(::grpc::ServerContext* context, const Read
-#else
   try {
-    ViObject object_handle = request->object_handle();
-    ViAttr attribute_name = request->attribute_name();
-    ViAttrState attribute_value = request->attribute_value();
-    auto status = library_->SetAttribute(object_handle, attribute_name, attribute_value);
-    if (!status_ok(status)) {
-      return ConvertApiErrorStatusForViObject(context, status, object_handle);
+    auto vi_grpc_session = request->vi();
+    ViSession vi = session_repository_->access_session(vi_grpc_session.name());
+    ViAttr attributeID = request->attribute_name();
+    ViAttrState attribute_value = 0;
+    if (request->attribute_value().has_value_bool()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_bool() ? VI_TRUE : VI_FALSE);
+    }
+#if defined(_VISA_ENV_IS_64_BIT)
+    else if (request->attribute_value().has_value_u64()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_u64());
     }
-    response->set_status(status);
-    return ::grpc::Status::OK;
-  }
-  catch (nidevice_grpc::NonDriverException& ex) {
-    return ex.GetStatus();
-  }
 #endif
-}
-//---------------------------------------------------------------------
-//---------------------------------------------------------------------
-::grpc::Status VisaService::StatusDesc(::grpc::ServerContext* context, const StatusDescRequest* request, StatusDescResponse* response)
-{
-  if (context->IsCancelled()) {
-    return ::grpc::Status::CANCELLED;
-  }
-#if 1
-  return ::grpc::Status(grpc::StatusCode::DO_NOT_USE, "Custom code not implemented yet");
-#else
-  try {
-    ViObject object_handle = request->object_handle();
-    ViStatus status_value = request->status_value();
-    std::string status_description(256 - 1, '\0');
-    auto status = library_->StatusDesc(object_handle, status_value, (ViChar*)status_description.data());
+    else if (request->attribute_value().has_value_i32()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_i32());
+    }
+    else if (request->attribute_value().has_value_u32()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_u32());
+    }
+    else if (request->attribute_value().has_value_i16()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_i16());
+    }
+    else if (request->attribute_value().has_value_u16()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_u16());
+    }
+    else if (request->attribute_value().has_value_u8()) {
+      attribute_value = static_cast<ViAttrState>(request->attribute_value().value_u8());
+    }
+    else {
+      throw nidevice_grpc::NonDriverException(::grpc::INVALID_ARGUMENT, "The value for attribute_value was not specified or out of range");
+    }
+    auto status = library_->SetAttribute(vi, attributeID, attribute_value);
```

Comment:
Are you OK with the behavior if the user-provided value does not match the type of the attribute ID? It looks like the worst case would be truncation or maybe roll over?

DAQ has a varargs generic accessor API where incorrect types could be a crash or other undefined behavior. For that we had a check in the service to make sure they matched and there's a config token to allow unknown attributes (i.e., potentially added in future versions). Although I think that's only valid for "raw" attribute IDs.

Reply:
The server is a 64-bit process, and `ViAttrState` is a 64-bit value. All the others will be smaller and will increase in size so they won't be truncated. If the user passes an attribute size that's smaller than needed - let's say a timeout of just over 4 seconds, 4096 specifically, and they pass it as i16 or u16 - that's no different than putting the constant 4096 on the stack. So I'm not worried here. And if they pass a *too-large* value, such as `0xFFFFFFFF00001000ULL`, well VISA is going to end up taking the 64-bit value and truncating it there. So I'm not really worried about that either.